Add DevGuide documentation of Brigand

wthrowe opened the pull request:
## Proposed changes

As briefly discussed in #2255, Brigand has almost no documentation and we should try to fix this.  This PR adds documentation to the DevGuide.  We can try to contribute back to Brigand once we are satisfied here.

I've put up a [rendered version](http://hosting.astro.cornell.edu/~wthrowe/brigand_docs/brigand.html) of the current state.  I'll try to remember to keep it up to date as I modify things.

### Types of changes:

- [ ] Bugfix
- [ ] New feature
- [ ] Refactor

### Component:

- [ ] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

wthrowe committed:
Add DevGuide documentation of Brigand

nilsvu commented:
Thanks for doing this @wthrowe! What I would find very useful is to "opinionate" the documentation, i.e. to note which of the `tmpl` metafunctions we prefer over the `std` equivalent, and why, along with other guidelines on using them. What do you think?

wthrowe commented:
Changed a bit.  (The link in the initial post is updated.)  The significant changes are:
* Added a few new functions.
* Adding some `\remark`s about similar STL functions when they exist.  Some of the preferences between the versions could still use more rationale.
* Experimented with an `assert_same` function template as an abbreviation for the repeated `static_assert(std::is_same_v<...>)` boilerplate.  It is currently demonstrated for `int32_t` and `int64_t`.  I think it makes things a bit more readable, so I'll probably switch to that if no one complains.

wthrowe commented:
And here's the much-awaited metalambda update!  People should read it and tell me what needs more explanations or examples or whatever.

I've updated the [rendered version](http://hosting.astro.cornell.edu/~wthrowe/brigand_docs/brigand.html) I posted earlier with the new content.

@nilsdeppe, could you check in particular the general background to metaprogramming I wrote and let me know if I've screwed anything up or completely missed the point or something?

wthrowe commented:
All of Brigand is now documented.  Still missing documentation for a few SpECTRE things and some necessary discussion.

kidder commented on docs/DevGuide/Brigand.md:
```diff
@@ -0,0 +1,1795 @@
+
+\par
+Most of the standard Brigand functions are eager, but many have lazy versions
+in the nested `lazy` namespace.  Thesw are indicated by calls to the
```
Thesw -> These

kidder commented on docs/DevGuide/Brigand.md:
```diff
@@ -0,0 +1,1795 @@
+when the lambda is used.  The lazy nature of the metafunctions prevents them
+from prematurely evaluating to results based on the literal placeholder types.
+The \ref apply function can be used to evaluate a metalambda with specified
+arguments, any many other Brigand functions take metalambdas that are evaluated
```
any many -> and many ?

kidder reviewed (changes requested):
a couple of typos on my first pass; I will do a more careful reading of the intro and metalambda sections next...

wthrowe committed:
Simplify implementation of remove_duplicates

wthrowe commented:
Finished documenting the "popular" SpECTRE extensions.  Also changed SpECTRE's implementation of remove_duplicates.

Forgot to fix the typos.  I'll get them in the next update.

nilsdeppe commented on docs/DevGuide/Brigand.md:
```diff
@@ -0,0 +1,1823 @@
+\subsubsection bind Bind expression
+
+\par
+A bind expression is a specialization of `bind`.  It wraps an eager
```
picking one of these at random: do you think it would be clearer or better to say "is a specialization of `tmpl::bind`"? That is, adding the `tmpl` in front. I'm undecided but I don't see how it hurts other than making more work, so I'm fine not doing it because this amount of documentation is amazing as-is :D

wthrowe replied:
That's probably clearer, yeah.  I'll go through and find the other places.  (Most things are links, and I think those are clear enough.)

nilsdeppe replied:
I agree, links are definitely clear enough :)

nilsdeppe commented on docs/DevGuide/Brigand.md:
```diff
@@ -0,0 +1,1823 @@
+\subsubsection multiplication_table
+
+\par
+Constructs a multiplication table.
```
This is amazing!!!! :D :D

nilsdeppe commented on docs/DevGuide/Brigand.md:
```diff
@@ -0,0 +1,1823 @@
+the branch is invalid because it would recurse infinitely.
+
+
+\subsubsection primes
```
This is some serious TMP... :+1:

nilsdeppe commented on docs/DevGuide/Brigand.md:
```diff
@@ -0,0 +1,1823 @@
+\snippet Test_TMPLDocumentation.cpp tmpl::type_
+
+\par
+When extracting the type, programmers are encouraged to use \ref type_from.
```
Maybe add "to make it clear that the `::type` that would otherwise appear is not an evaluation of a lazy metafunction."
I guess technically it is, but `type_from` is mostly to give context, I find.

wthrowe replied:
That's a good way to explain it.

If you want a metafunction, you should use `always` or `identity` instead, which behave the same way.  (Brigand has a lot of functionality duplication.)

nilsdeppe commented on docs/DevGuide/Brigand.md:
```diff
@@ -0,0 +1,1823 @@
+\par
+
+\note
+An object of the list template parameter type is constructed, so the list must
```
Does this mean the elements of the list must all be complete types? Looking at the Brigand implementation, I think if instead `for_each_impl` took a `List<Elements...>*` would that fix the need for having complete types? Otherwise, I think using a `struct` partial specialization could also fix the issue, right?

This is mostly for my understanding :)

wthrowe replied:
Ah, no, the list itself is constructed (`tmpl::list<>{}`).  This is irrelevant if you are using a `tmpl::list`, but might be unpleasant if you pass in something like a TaggedTuple.

I finally got around to pushing my patches to my fork.  I just solved this one using a `type_`: wthrowe/brigand@9872c8f7479fc8a7cc17e61fa1ef074051264960.

nilsdeppe replied:
ooh sweet! :D Now if only we could get Brigand folks to respond on your PR -_-

nilsdeppe commented on tests/Unit/Utilities/Test_TMPLDocumentation.cpp:
```diff
@@ -0,0 +1,1467 @@
+template <typename... T>
+bool check_sizes(const T&... containers,
+                 const typename tmpl::has_type<T, size_t>::type... sizes) {
+  return (... && (containers.size() == sizes));
```
Yay fold expressions :D Minor comment, could `and` be used instead of `&&` and would that be reasonable to do?

wthrowe replied:
I didn't think to try.  It probably works, since `and` is just token replacement.

nilsdeppe reviewed (commented):
This is amazing! Thank you so much for putting all this documentation together. I really appreciate it :D I just have a few minor comments, none of which are binding, just things for you to consider. I'm happy for you to squash everything (including any new changes). I'm happy for this to be merged once squashed :)

markscheel commented on docs/DevGuide/Brigand.md:
```diff
@@ -0,0 +1,1827 @@
+\subsection lazy Eager and lazy metafunctions
+
+\par
+Metafunctions commonly appear in two forms: eager and lazy.  Lazy functions are
```
Lazy functions -> Lazy metafunctions

markscheel commented on tests/Unit/Utilities/Test_TMPLDocumentation.cpp:
```diff
@@ -0,0 +1,1467 @@
+namespace metafunctions {
+/// [metafunctions:lazy]
+template <typename T>
+struct lazy_add_list {
```
Should `lazy_add_list` be renamed `lazy_make_list_of_T_and_int` or something like that, since it doesn't add anything to an existing list?  Same for the eager version.

wthrowe replied:
Yes.  I don't know how I got the current name.

markscheel commented on tests/Unit/Utilities/Test_TMPLDocumentation.cpp:
```diff
@@ -0,0 +1,1467 @@
+  using type = List1<T...>;
+};
+
+// Type1 < Type2 < Type3
```
What does the 'less than' operator mean in this comment?  The metalambda seems to compare only type equality or inequality.  And the metalambda doesn't refer to Type2.

wthrowe replied:
That's supposed to be the definition of `<`.  I'll make that clearer.

markscheel commented on tests/Unit/Utilities/Test_TMPLDocumentation.cpp:
```diff
@@ -0,0 +1,1467 @@
+template <template <typename> typename Func, typename DummyArg, typename Arg>
+struct eager_call_func_impl<Func<DummyArg>, Arg> {
+  using type = Func<Arg>;
+};
```
I think this needs a few more words and renaming of structs to make it less confusing.   The first thing that the reader sees is that `eager_call_func_impl` is implemented differently than `lazy_call_func_impl` above, and this makes the reader think that the eager vs lazy being demonstrated here is the different implementation of the `call_func_impl`s.  And then the reader gets confused because `eager_call_func_impl` is itself a lazy metafunction but it starts with the word "eager". But (I think) the real point is that `eager_add_list` evaluates its dummy argument immediately, so it cannot be used as a functor.

So I suggest:
 - Please rename `eager_call_func_impl` to `call_eager_metafunction_impl` , rename `eager_call_func` to `call_eager_metafunction` so that it is clear that the "eager" refers to the eagerness/laziness of `Func`, the metafunction you want to call, not the eagerness/laziness of `call_function_impl`.  And similar renaming for the lazy versions.
 - Please add some wording (in Brigand.md where this snippet is referenced) along the lines of "Note that the implementation of `call_eager_metafunction_impl` is different than that of `call_lazy_metafunction_impl`, and both are implemented in a lazy manner, but the implementation details are not the important point; the point is that `eager_add_list` evaluates its dummy argument immediately, so it cannot be used as a functor in the same way as `lazy_add_list`.

wthrowe replied:
You have correctly figured out what I was trying to get across here.  I will try to make it easier for the next person who has to do so.

markscheel commented on docs/DevGuide/Brigand.md:
```diff
@@ -0,0 +1,1827 @@
+\snippet Test_TMPLDocumentation.cpp tmpl::pin
+
+\subsubsection defer Defer expression
+
```
I think `defer` needs more explanation and a better/simpler example.  In particular, `tmpl::always` is used in the example but hasn't been explained yet.  Also, in the example, `Type1` appears to have just been thrown away, but presumably it has been captured (and then later thrown away).  I do like the example `maybe_first` for how to delay evaluation of something that will fail if evaluated early.

wthrowe replied:
The always was a workaround for a bug.  The bug is fixed now, so I will remove it.  I can't simplify the example more than that, though, except for possibly dropping the `Type1`.

The `Type1` is for similarity with the next example, but what is going on should be made clearer.

markscheel commented on docs/DevGuide/Brigand.md:
```diff
@@ -0,0 +1,1827 @@
+
+
+\subsubsection maybe_first
+
```
This seems to be the simplest example that uses `tmpl::defer`.  Maybe add an explanation for why you need the `tmpl::pin`?  And maybe an explanation of what is returned by the inner `apply` and what is returned by the outer `apply`?

wthrowe replied:
I should write a lot more explanation of what is going on in all of these examples.

markscheel replied:
Yes, I like the new explanation much better!

markscheel commented on docs/DevGuide/Brigand.md:
```diff
@@ -0,0 +1,1827 @@
+If another type is needed, it can be wrapped in \ref always.
+
+\subsubsection parent Parent expression
+
```
Same issue with `parent` as with `defer`.  Would be nice to have a better simple example of a metaclosure.  Is there an example that is analogous to python's
```
def f(x):
   return lambda y: x + y
a = f(2)
print(a(3)) # prints 5
```
where both the captured value and some function argument are both used together?

wthrowe replied:
This will make a good addition to the early simple part of the examples section.

markscheel commented on docs/DevGuide/Brigand.md:
```diff
@@ -0,0 +1,1827 @@
+and, possibly, a parent context.  When evaluating a metalambda, for example by
+calling \ref apply, the initial context contains the passed set of arguments
+with no parent context.
+
```
If `apply` has an initial context, does it also have a final context? If so, how does its context change?  Also, I don't understand 'parent' in this usage; is 'parent' like a base class? I think contexts could be made clearer by giving examples of contexts.

wthrowe replied:
I wrote this thinking of a sort of "procedural" evaluation where you are examining and evaluating things in some sequence, rather than a "functional" one where you, for example, make function calls and then wait for results to be reported to you.  As a c example, consider running
```c
int main(int argc, char** argv) {
  return sin(2);
}
```
(No that doesn't make any sense.)  In the "proocedural" sense, the initial context is "in main, arguments argc, argv", but later it will be "in sin, argument 2".  In the "functional" sense, you have the context of main (args argc, argv) and the context of sin (arg 2), but there's not really a "current" context.

This is a functional language, so thinking the latter way is probably better.  A fully worked example will hopefully make this all clearer.

---

A parent is like the next entry in a singly linked list.  I need to add better discussion of contexts outside the descriptions of the functions that manipulate them.

markscheel replied:
I like the explanations in terms of a stack much better than in terms of contexts!

markscheel commented on docs/DevGuide/Brigand.md:
```diff
@@ -0,0 +1,1827 @@
+An argument is one of the structs `tmpl::_1`, `tmpl::_2`, or `tmpl::args<N>`
+for `unsigned int` N.  The additional aliases `tmpl::_3`, `tmpl::_4`, ...,
+`tmpl::_9` are provided to `tmpl::args<2>`, `tmpl::args<3>`, ...,
+`tmpl::args<8>`.
```
Just out of curiosity, is there a reason that `tmpl::_1` is not aliased to `tmpl::args<0>`? Is it just to avoid confusion by preventing people from using start-counting-at-zero and start-counting-at-one arguments in the same metafunction?

wthrowe replied:
I have not the foggiest idea.  There are several places where there are checks for `tmpl::_1` and sometimes `tmpl::_2` as common optimized cases, but I don't think there's a reason that couldn't be done with `tmpl::args<0>`.

markscheel commented on tests/Unit/Utilities/Test_TMPLDocumentation.cpp:
```diff
@@ -0,0 +1,1467 @@
+assert_same<tmpl::apply<tmpl::pin<List1<Type1>>>, List1<Type1>>();
+// Error: List1 is not a lazy metafunction
+// assert_same<tmpl::apply<List1<Type1>>, List1<Type1>>();
+assert_same<tmpl::apply<tmpl::pin<tmpl::_1>, Type1>, tmpl::_1>();
```
This last example is interesting, and highlights a potential confusion.  In the accompanying section in Brigand.md, please point out that `tmpl::pin<tmpl::_1>` is interpreted as a metafunction taking one argument, but `tmpl::pin<List1<Type1>>` is interpreted as a metafunction taking zero arguments, even though in both cases `tmpl::pin` has a single template argument.  Unlike most languages, where the number of function arguments is determined only by position (that is `f(x,y)` is a function taking two arguments no matter what you rename x and y), here the number of arguments depends on the name of the argument (i.e. whether it contains the special tokens `tmpl::_1` and so on).

wthrowe replied:
Something isn't clear here, I guess.  Both `tmpl::pin<tmpl::_1>` and `tmpl::pin<List1<Type1>>` are constants (equal to `tmpl::_1` and `List1<Type1>`, respectively).  As metalambdas, brigand treats constants as function that evaluate to themselves.

The general observation is correct, though.  The number of arguments a function requires can't be determined without examining the entire body.  (And you are free to pass extra arguments, which will be ignored.)

markscheel commented on docs/DevGuide/Brigand.md:
```diff
@@ -0,0 +1,1827 @@
+\par
+Metalambdas must be passed enough arguments to define all argument placeholders
+in their bodies.  Failure to pass enough arguments may error or produce
+unintuitive results.
```
Possible rewrite explaining how to pass arguments:

When evaluating a metalambda, the metalambda must be passed enough arguments to define all argument placeholders in their bodies. Arguments are passed by appending them to the template parameters of `tmpl::apply`. Failure to pass enough arguments may error or may produce unintuitive results.

markscheel commented on docs/DevGuide/Brigand.md:
```diff
@@ -0,0 +1,1827 @@
+\snippet Test_TMPLDocumentation.cpp tmpl::parent
+
+\warning
+Do not call `tmpl::parent` outside of a metaclosure context.  This results in
```
'outside of a metaclosure context' means 'outside of a `tmpl::defer`'?

wthrowe replied:
Mostly.  All metaclosure contexts are inside of `defer`s, but not everything in a `defer` is necessarily in a metaclosure context.  Basically, the rule is that you must always have "number of enclosing `defer`s >= number of enclosing `parent`s".

markscheel commented on docs/DevGuide/Brigand.md:
```diff
@@ -0,0 +1,1827 @@
+Do not call `tmpl::parent` outside of a metaclosure context.  This results in
+an empty evaluation context, causing unintuitive changes to the evaluation
+rules.  (Most, but not all, expressions are left unevaluated in such a
+context.)  Use \ref pin "pin" to prevent evaluation.
```
Please rewrite the last sentence as "If you are outside a metaclosure and you want to prevent evaluation of a metafunction, use pin instead of parent." (if I am interpreting this correctly)

markscheel commented on docs/DevGuide/Brigand.md:
```diff
@@ -0,0 +1,1827 @@
+argument, replacing the current context with its parent.  This provides access
+to the captured arguments in a metaclosure.
+\snippet Test_TMPLDocumentation.cpp tmpl::parent
+
```
I guess I don't have a picture in my mind for why this is called `parent`. Usually when I think of a "parent" I think of some hierarchical structure like a filesystem or an inheritance tree or a linked list that you can visualize as a tree. Are parents supposed to be for metafunctions of metafunctions or some similar nested structure?

wthrowe replied:
I tried to describe it as a linked list above, but it might be easier to think of the structures as a stack.  The variables at the top of the stack are what you get from `tmpl::_1`, etc.  `parent` pops the top set of values off the stack.  `defer` copies the stack into a metaclosure.  Evaluating a metaclosure replaces the current stack with the one from the metaclosure, but also pushes the variables at the top of the old stack onto the new stack.

This might be easier to explain if I went into an explanation at this abstract level before getting to the actual brigand functions.  Perhaps putting all the closure-related functions after a section explaining all this "context" stuff would be clearer.

Diagrams might also help.

markscheel commented on docs/DevGuide/Brigand.md:
```diff
@@ -0,0 +1,1827 @@
+\subsubsection metalambda_metaclosure Metaclosure
+
+\par
+A metaclosure is an opaque type produced by \ref defer "defer", containing a
```
What is an "opaque type"?

wthrowe replied:
https://en.wiktionary.org/wiki/opaque:
5. (computing) Describes a type for which higher-level callers have no knowledge of data values or their representations; all operations are carried out by the type's defined abstract operators.

Basically an object that you cannot manipulate or obtain information from directly.  A c++ lambda type is also one: all you can do is call it, and there are no guarantees about what it stores internally.

markscheel commented on docs/DevGuide/Brigand.md:
```diff
@@ -0,0 +1,1827 @@
+evaluates the packaged metalambda in the current evaluation context with the
+parent context replaced by the packaged context.  See \ref defer and \ref
+parent for examples.
+
```
Here again it would be nice to see a concrete example stepped through, in order to demonstrate what contexts are. e.g. "In this example, after the first `apply` the context is bla, and after the second `apply` the context is bla2".

markscheel commented on docs/DevGuide/Brigand.md:
```diff
@@ -0,0 +1,1827 @@
+\snippet Test_TMPLDocumentation.cpp tmpl::empty_sequence
+
+\remark
+Prefer just writing an empty list.
```
"Prefer just writing `tmpl::list<>`".

markscheel commented on docs/DevGuide/Brigand.md:
```diff
@@ -0,0 +1,1827 @@
+\snippet Test_TMPLDocumentation.cpp tmpl::type_
+
+\par
+When extracting the type, programmers are encouraged to use \ref type_from to
+make it clear that the `::type` that would otherwise appear is not an
+evaluation of a lazy metafunction.  See \ref always or \ref identity for
+similar functionality that is intended for use as a metafunction.
```
Do you mean

"When extracting the type, programmers are encouraged to use `tmpl::type_from<tmpl::type_<T>>` instead of `tmpl::type_<T>::type` because the latter looks like the evaluation of a lazy metafunction, which might cause confusion." ?

Also, is one supposed to prefer `tmpl::always` or `tmpl::identity`?

wthrowe replied:
Yes.

Brigand classifies `identity` as a mathematical operator, so I guess you could go with that as a rule.  From a practical standpoint, `always`, `identity`, and `type_from` all behave identically.

markscheel commented on docs/DevGuide/Brigand.md:
```diff
@@ -0,0 +1,1827 @@
+\snippet Test_TMPLDocumentation.cpp tmpl::false_type
+
+\remark
+Prefer std::false_type.
```
Is `tmpl::false_type` interchangable with `std::false_type`?   (i.e. is it always possible to use `std::false_type` instead of `tmpl::false_type`?)

wthrowe replied:
Except for the fact that they are different types and so will act differently when something cares about type equality, the `std` version is a drop-in replacement for the `tmpl` version.

markscheel commented on docs/DevGuide/Brigand.md:
```diff
@@ -0,0 +1,1827 @@
+\snippet Test_TMPLDocumentation.cpp tmpl::integral_list
+
+\remark
+Prefer std::integer_sequence when used for pack expansion.  Prefer this when
```
this -> integral_list

markscheel commented on tests/Unit/Utilities/Test_TMPLDocumentation.cpp:
```diff
@@ -0,0 +1,1467 @@
+
+/// [tmpl::count_if:bug:asserts]
+assert_same<tmpl::count_if<List1<Type1>, ReturnTrueFalse<void>>,
+            tmpl::integral_constant<size_t, 0>>();
```
Add comment // should return tmpl::integral_constant<size_t, 1>

wthrowe replied:
The documented bugs have been fixed, so these sections will be removed.

markscheel commented on docs/DevGuide/Brigand.md:
```diff
@@ -0,0 +1,1827 @@
+
+\par
+Checks whether a particular type is contained in a list, returning an
+\ref integral_constant of `bool`.
```
When should one choose to return `tmpl::integral_constant<bool,...>` versus `tmpl::true_type`/`tmpl::false_type` versus `std::true_type/std::false_type`?

wthrowe replied:
My recommendations:
* If you are writing a metafunction implemented by returning the result of another metafunction, don't do any conversion but directly return whichever of `std` or `tmpl` the wrapped thing returns.
* If you are creating boolean constants yourself:
  * Use the `std` versions unless comparing type equality with the result of something that returns `tmpl` versions.
  * Use `true_type` or `false_type` if you want a literal true or false, and `integral_constant` (or `std::bool_constant` or `tmpl::bool_`) if you want to create it from an expression.

This particular function is implemented in SpECTRE but falls under the first bullet.

markscheel commented on tests/Unit/Utilities/Test_TMPLDocumentation.cpp:
```diff
@@ -0,0 +1,1467 @@
+assert_same<tmpl::apply<tmpl::lazy::replace<tmpl::pin<List1<Type1>>,
+                                            tmpl::_1, Type2>,
+                        Type1>,
+            List1<Type1>>();
```
Pls add comment // should be List1<Type2>

markscheel commented on docs/DevGuide/Brigand.md:
```diff
@@ -0,0 +1,1827 @@
+\snippet Test_TMPLDocumentation.cpp tmpl::eval_if
+
+\par
+This is performs lazy evaluation of conditional branches outside of a
```
This is -> This

markscheel reviewed (changes requested):
Thanks @wthrowe for doing this!

Going through all the bugs and limitations of a huge number of brigand metafunctions is a lot of work!

I am also very happy to see the beginning sections which explain the concepts behind everything so that it is easier to understand what the individual brigand metafunctions do, and I think this is a very important part of the documentation.  So I put most of my comments there.

My overall comment on the first sections is that it would be nice to add more explanation of what 'contexts' are and how metaclosures work, with really simple examples that are stepped through and explained at each stage of evaluation.  I'm glad that there is some explanation in the later examples (which would make Eratosthenes proud, are great to have so one can see what is possible, but are daunting for a beginner), but I think more exposition is needed in the earlier, simpler examples.

wthrowe reviewed (commented):
Thanks for the review.  I think I see how I can make the discussions around metaclosures clearer.

markscheel commented on docs/DevGuide/Brigand.md:
```diff
@@ -0,0 +1,2014 @@
+\snippet Test_TMPLDocumentation.cpp metafunctions:call_lazy_metafunction
+and get the expected output:
+\snippet Test_TMPLDocumentation.cpp metafunctions:call_lazy_metafunction_assert
+A similar attempt with an eager metafunction fails, because the function is
```
"A similar attempt with an eager metafunction fails..." -> "But it fails if you try to call an arbitrary *eager* metafunction twice in the same way..."

(trying to make it clearer that eager/lazy refers to `Func` and not to `apply...`)

markscheel commented on docs/DevGuide/Brigand.md:
```diff
@@ -0,0 +1,2014 @@
+parameters is evaluated as a metalambda and replaced by the result, and then
+the struct's `type` type alias is the result of the full lazy-expression.
+\snippet Test_TMPLDocumentation.cpp metalambda_lazy
+
```
So the lazy expression here is `make_list1` ?

wthrowe replied:
Renamed `make_list1` to `lazy_make_list1` so people don't have to look up what it is to understand the example.  Does that clarify it enough?

markscheel commented on docs/DevGuide/Brigand.md:
```diff
@@ -0,0 +1,2014 @@
+
+\par
+A parent expression is a specialization of `tmpl::parent`.  It evaluates its
+argument (treated as a metalambda), evaluated after popping the last entry off
```
"popping the last entry" -> "popping the first entry"?  or "popping the top entry"?  or "popping the front entry"? or "popping an entry"?

("last" sounds like the bottom of the stack, like `std::deque.back()`.  I think we want language like `std::stack::pop()`.)

markscheel commented on docs/DevGuide/Brigand.md:
```diff
@@ -0,0 +1,2014 @@
+4. The \ref metalambda_metaclosure "metaclosure" evaluates the contained
+   `tmpl::parent<tmpl::_1>` with the argument stack [(`Type2`), (`Type1`)].
+
+5. Evaluating the `tmpl::parent` replaces pops the stack, and so evaluates
```
"replaces pops" -> "pops"

markscheel commented on docs/DevGuide/Brigand.md:
```diff
@@ -0,0 +1,2014 @@
+argument passed to it (`tmpl::_1`, which is 5 in the example) and the value
+captured at it's creation (`tmpl::parent<tmpl::_1>`, which is 3 in the
+example).
+
```
Yes, good example!

markscheel reviewed (commented):
A few small comments.  But very much improved!  I like the worked-out examples and the improved explanation of metaclosures, defer, parent, etc.

wthrowe reviewed (commented):
Updated for @markscheel's comments (and rebased on develop).

markscheel reviewed (commented):
Looks good to me! Please squash.

knelli2 commented on docs/DevGuide/Brigand.md:
```diff
@@ -0,0 +1,2015 @@
+resulting in the `List1` metafunction being acted upon instead of
+`eager_add_list`:
+\snippet Test_TMPLDocumentation.cpp metafunctions:call_eager_metafunction
+\snippet Test_TMPLDocumentation.cpp metafunctions:call_eager_metafunction_assert
+(In this simple case we could have used a template template parameter to pass
```
Thought `template template` was a typo the first time I saw it. Maybe add a comment that this isn't a typo and refers to `template<template<...>>`

wthrowe replied:
I added a link to cppreference.

knelli2 reviewed (commented):
Thanks for doing this! The early versions were very helpful when I was starting out and I still use it now 😄

wthrowe reviewed (commented):
Squashed everything through the changes for @markscheel's most recent comments.

Added fixups for @knelli2's comments and a few minor things I thought of myself.

knelli2 commented on docs/Installation/Installation.md:
```diff
@@ -61,7 +61,7 @@ all of these dependencies.
 * BLAS (e.g. [OpenBLAS](http://www.openblas.net))
 * [Blaze](https://bitbucket.org/blaze-lib/blaze/overview) v3.8
 * [Boost](http://www.boost.org/) 1.60.0 or later
-* [Brigand](https://github.com/edouarda/brigand)
+* [Brigand](https://github.com/edouarda/brigand) from 4 October 2020 or later
```
(Optional) Giving a date that isn't an official release might make it confusing which version to use. Maybe give a commit hash instead?

wthrowe replied:
Good idea.  Changed to "at commit 1c398e4f1e817ab195e4cd6fbb03c18cb386eea3 (late 2020) or later".  I decided to retain a vague date so people with a much newer download don't have to look through the repo to verify they are OK.

knelli2 reviewed (commented):
Looks good! You can squash in the optional change if you want or not. Up to you :)

nilsvu commented:
@wthrowe are you ready to get this merged? Then please mark the PR as "Ready for review".

wthrowe commented:
Thank you, yes.

nilsvu commented:
@markscheel and @kidder your requested changes still show up as unresolved. Please approve if you agree to merge. I suggest to merge this PR and do further changes in follow-ups.
